File: Omatic.Deploy/FixHtmlFormsPaths.vb
Imports Microsoft.Build.Framework
Imports Microsoft.Build.Utilities
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions

Namespace Omatic.Deploy
    Public Class FixHtmlFormsPaths
        Inherits Task
        Private _fixedPaths As List(Of ITaskItem)

        <Required()>
        Public Property Paths As ITaskItem()

        <Required()>
        Public Property DestinationFolder As String

        <Output()>
        Public ReadOnly Property FixedPaths() As ITaskItem()
            Get
                Return _fixedPaths.ToArray()
            End Get
        End Property

        Public Overrides Function Execute() As Boolean
            _fixedPaths = New List(Of ITaskItem)()

            Log.LogMessage("Fixing HTML forms file paths")

            Dim remove As New Regex("^.*\\htmlforms\\", RegexOptions.IgnoreCase Or RegexOptions.Compiled)
            For Each path As String In Paths.[Select](Function(x) x.ItemSpec)
                Dim newPath As String = System.IO.Path.Combine(DestinationFolder, remove.Replace(path, ""))
                Log.LogMessage(MessageImportance.Low, "Old path: ""{0}""" & vbTab & "New path: ""{1}""", path, newPath)
                _fixedPaths.Add(New TaskItem(newPath))
            Next
            Return True
        End Function
    End Class
End Namespace

File: OmaticSoftware.Deploy/FixHtmlFormsPaths.cs
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text.RegularExpressions;

namespace OmaticSoftware.Deploy
{
    public class FixHtmlFormsPaths : Task
    {

        private List<ITaskItem> _fixedPaths;
        [Required()]
        public ITaskItem[] Paths { get; set; }

        [Required()]
        public string DestinationFolder { get; set; }

        [Output()]
        public ITaskItem[] FixedPaths
        {
            get { return _fixedPaths.ToArray(); }
        }

        public override bool Execute()
        {
            _fixedPaths = new List<ITaskItem>();

            Log.LogMessage("Fixing HTML forms file paths");

            Regex @remove = new Regex("^.*\\\\htmlforms\\\\", RegexOptions.IgnoreCase | RegexOptions.Compiled);
            foreach (string path in Paths.Select(x => x.ItemSpec))
            {
                string newPath = System.IO.Path.Combine(DestinationFolder, @remove.Replace(path, ""));
                Log.LogMessage(MessageImportance.Low, "Old path: \"{0}\"\tNew path: \"{1}\"", path, newPath);
                _fixedPaths.Add(new TaskItem(newPath));
            }
            return true;
        }
    }
}
